warehouse/macaroons/models.py

```diff
@@ -37,15 +38,29 @@ class Macaroon(db.Model):
         UniqueConstraint(
             "description", "user_id", name="_user_macaroons_description_uc"
         ),
+        UniqueConstraint(
+            "description", "project_id", name="_project_macaroons_description_uc"
+        ),
+        CheckConstraint(
+            "(user_id::text IS NULL) <> (project_id::text IS NULL)",
+            name="_user_xor_project_macaroon",
+        ),
```

Comment:
cc @di: This enforces the XOR of "every Macaroon is either user-associated or project-associated." Let me know what you think about this approach.